src/catalog/catalog.cpp

```diff
@@ -30,413 +30,764 @@ Catalog *Catalog::GetInstance(void) {
-      LOG_TRACE("Database already exists. Returning ResultType::FAILURE.");
-      return ResultType::FAILURE;
-    }
+  {
+    std::lock_guard<std::mutex> lock(catalog_mutex);
+    databases_.push_back(database);
   }
-  oid_t database_id = GetNextOid();
-  storage::Database *database = new storage::Database(database_id);
-  database->setDBName(database_name);
-  databases_.push_back(database);
 
-  InsertDatabaseIntoCatalogDatabase(database_id, database_name, txn);
+  // Insert database tuple
+  pg_database->InsertDatabase(database_oid, database_name, pool_.get(), txn);
 
   LOG_TRACE("Database created. Returning RESULT_SUCCESS.");
   return ResultType::SUCCESS;
 }
 
-void Catalog::AddDatabase(storage::Database *database) {
-  databases_.push_back(database);
-  std::string database_name;
-  InsertDatabaseIntoCatalogDatabase(database->GetOid(), database_name, nullptr);
-}
-
-void Catalog::AddDatabase(std::string database_name,
-                          storage::Database *database){
-  database->setDBName(database_name);
-  databases_.push_back(database);
-  LOG_DEBUG("Added database with name: %s", database_name.c_str());
-  InsertDatabaseIntoCatalogDatabase(database->GetOid(), database_name, nullptr);
-}
-
+ResultType Catalog::CreateTable(const std::string &database_name,
+                                const std::string &table_name,
+                                std::unique_ptr<catalog::Schema> schema,
+                                concurrency::Transaction *txn) {
+  auto &txn_manager = concurrency::TransactionManagerFactory::GetInstance();
+  bool single_statement_txn = false;
 
-void Catalog::InsertDatabaseIntoCatalogDatabase(oid_t database_id,
-    std::string &database_name, concurrency::Transaction *txn) {
-  // Update catalog_db with this database info
-  auto tuple =
-      GetDatabaseCatalogTuple(
-          databases_[START_OID]->GetTableWithName(DATABASE_CATALOG_NAME)->GetSchema(),
-          database_id, database_name, pool_);
-  catalog::InsertTuple(
-      databases_[START_OID]->GetTableWithName(DATABASE_CATALOG_NAME),
-      std::move(tuple), txn);
-}
+  if (txn == nullptr) {
+    single_statement_txn = true;
+    txn = txn_manager.BeginTransaction();
+  }
 
-// Create a table in a database
-ResultType Catalog::CreateTable(std::string database_name, std::string table_name,
-    std::unique_ptr<catalog::Schema> schema, concurrency::Transaction *txn) {
   LOG_TRACE("Creating table %s in database %s", table_name.c_str(),
-      database_name.c_str());
+            database_name.c_str());
+
+  oid_t database_oid =
+      DatabaseCatalog::GetInstance()->GetDatabaseOid(database_name, txn);
+  if (database_oid == INVALID_OID) {
+    LOG_TRACE(
+        "Cannot find the database. Return "
+        "RESULT_FAILURE.");
+    return ResultType::FAILURE;
+  }
+
+  oid_t table_oid =
+      TableCatalog::GetInstance()->GetTableOid(table_name, database_oid, txn);
+  if (table_oid != INVALID_OID) {
+    LOG_TRACE(
```

Comment:
I'm not sure about the differences between throw catalog exceptions and set transaction's result type as failure. 
e.g. when user try to create a table with duplicate column names, catalog will throw exception or just set transaction result as failure?